rllib/execution/rollout_ops.py

```diff
@@ -112,31 +122,52 @@ def synchronous_parallel_sample(
                         "No healthy remote workers left. Trying to restore workers ..."
                     )
                 break
+
+            if _return_metrics:
+                sampled_data = [s[0] for s in sampled_data]
+                stats_dicts = [s[1] for s in sampled_data]
+
         # Update our counters for the stopping criterion of the while loop.
-        for batch_or_episode in sampled_data:
+        if _return_metrics:
             if max_agent_steps:
-                agent_or_env_steps += (
-                    sum(e.agent_steps() for e in batch_or_episode)
-                    if _uses_new_env_runners
-                    else batch_or_episode.agent_steps()
+                agent_or_env_steps += sum(
+                    int(agent_stat)
+                    for stat_dict in stats_dicts
+                    for agent_stat in stat_dict[NUM_AGENT_STEPS_SAMPLED].values()
                 )
             else:
-                agent_or_env_steps += (
-                    sum(e.env_steps() for e in batch_or_episode)
-                    if _uses_new_env_runners
-                    else batch_or_episode.env_steps()
+                agent_or_env_steps += sum(
+                    int(stat_dict[NUM_ENV_STEPS_SAMPLED])
+                    for stat_dict in stats_dicts
                 )
+        else:
+            for batch_or_episode in sampled_data:
+                if max_agent_steps:
+                    agent_or_env_steps += (
```

Comment:
Do we need this, if we return metrics? ANd when don't we want to retrun metrics?

Reply:
Yeah, unfortunately, we do here in this case, b/c we have to determine right here (before even logging the actual steps to the Algorithm's metrics) when to stop the while loop.

We might rearrange this entire utility, but for now, it works just fine.